File: desktop/src/core/dependency_service.py
import os
import re
from typing import Set


class DependencyService:
    PYTHON_IMPORT_REGEX = re.compile(r'^\s*import\s+([a-zA-Z0-9_.,\s]+)', re.MULTILINE)
    PYTHON_FROM_REGEX = re.compile(r'^\s*from\s+(\.?\.?[a-zA-Z0-9_.]+)\s+import\s+([a-zA-Z0-9_.,\s()]+)', re.MULTILINE)

    JS_TS_ES6_REGEX = re.compile(r'\bimport\s+[^"\']*?\s+from\s+["\'](\..*?)["\']', re.DOTALL)
    JS_TS_DYNAMIC_REGEX = re.compile(r'\bimport\(\s*["\'](\..*?)["\']\s*\)')
    JS_TS_REQUIRE_REGEX = re.compile(r'\brequire\(\s*["\'](\..*?)["\']\s*\)')

    CPP_INCLUDE_REGEX = re.compile(r'^\s*#include\s+["\']([^"\']+)["\']', re.MULTILINE)

    RUST_MOD_REGEX = re.compile(r'^\s*(?:pub\s+)?mod\s+([a-zA-Z0-9_]+);', re.MULTILINE)
    RUST_USE_REGEX = re.compile(r'^\s*(?:pub\s+)?use\s+(crate|super|self)::([a-zA-Z0-9_:]+)', re.MULTILINE)

    def trace_dependencies(self, root_dir: str, target_rel_path: str) -> Set[str]:
        if not root_dir or not target_rel_path:
            return set()

        full_target_path = os.path.join(root_dir, target_rel_path)
        if not os.path.exists(full_target_path) or os.path.isdir(full_target_path):
            return set()

        try:
            with open(full_target_path, 'r', encoding='utf-8', errors='replace') as f:
                content = f.read()
        except Exception:
            return set()

        _, ext = os.path.splitext(target_rel_path)
        ext = ext.lower()
        found_rel_paths = set()
        target_dir = os.path.dirname(full_target_path)

        def check_and_add(path: str) -> bool:
            if os.path.isfile(path):
                rel = os.path.relpath(path, root_dir).replace('\\', '/')
                found_rel_paths.add(rel)
                return True
            return False

        if ext == '.py':
            targets = []

            for m in self.PYTHON_IMPORT_REGEX.finditer(content):
                for part in m.group(1).split(','):
                    clean = part.strip()
                    if clean:
                        targets.append(("", clean))

            for m in self.PYTHON_FROM_REGEX.finditer(content):
                base = m.group(1)
                clean_imports = m.group(2).replace('(', '').replace(')', '').replace('\n', ' ').replace('\r', ' ')
                for part in clean_imports.split(','):
                    clean_item = part.strip()
                    if clean_item:
                        targets.append((base, clean_item))

            for base, item in targets:
                has_base = bool(base)
                full_import_path = f"{base}.{item}" if has_base else item
                dot_count = len(base) - len(base.lstrip('.')) if has_base else len(item) - len(item.lstrip('.'))

                clean_import_path = full_import_path.lstrip('.')
                subpath = clean_import_path.replace('.', '/')

                if dot_count > 0:
                    temp_base = target_dir
                    for _ in range(dot_count - 1):
                        temp_base = os.path.dirname(temp_base)
                    search_bases = [temp_base]
                else:
                    search_bases = [
                        root_dir,
                        os.path.join(root_dir, 'src'),
                        target_dir
                    ]

                for base_dir in search_bases:
                    if subpath:
                        cand1 = os.path.join(base_dir, f"{subpath}.py")
                        cand2 = os.path.join(base_dir, subpath, "__init__.py")
                        if check_and_add(cand1) or check_and_add(cand2):
                            break
                    if has_base:
                        base_subpath = base.lstrip('.').replace('.', '/')
                        if base_subpath:
                            cand1 = os.path.join(base_dir, f"{base_subpath}.py")
                            cand2 = os.path.join(base_dir, base_subpath, "__init__.py")
                            if check_and_add(cand1) or check_and_add(cand2):
                                break

        elif ext in ('.js', '.jsx', '.ts', '.tsx'):
            imports = []
            for m in self.JS_TS_ES6_REGEX.finditer(content):
                imports.append(m.group(1))
            for m in self.JS_TS_DYNAMIC_REGEX.finditer(content):
                imports.append(m.group(1))
            for m in self.JS_TS_REQUIRE_REGEX.finditer(content):
                imports.append(m.group(1))

            for rel_import in imports:
                abs_base = os.path.normpath(os.path.join(target_dir, rel_import))
                possible_exts = ['', '.js', '.ts', '.jsx', '.tsx', '/index.js', '/index.ts']

                for p_ext in possible_exts:
                    test_path = abs_base + p_ext
                    if check_and_add(test_path):
                        break

        elif ext in ('.c', '.cpp', '.h', '.hpp'):
            for m in self.CPP_INCLUDE_REGEX.finditer(content):
                header = m.group(1)
                test_path_1 = os.path.join(target_dir, header)
                test_path_2 = os.path.join(root_dir, header)

                if not check_and_add(test_path_1):
                    check_and_add(test_path_2)

        elif ext == '.rs':
            for m in self.RUST_MOD_REGEX.finditer(content):
                mod_name = m.group(1)
                cand1 = os.path.join(target_dir, f"{mod_name}.rs")
                cand2 = os.path.join(target_dir, mod_name, "mod.rs")

                if not check_and_add(cand1):
                    check_and_add(cand2)

            for m in self.RUST_USE_REGEX.finditer(content):
                base_type = m.group(1)
                subpath = m.group(2).replace('::', '/')

                if base_type == 'crate':
                    src_dir = os.path.join(root_dir, 'src')
                    base_dir = src_dir if os.path.isdir(src_dir) else root_dir
                elif base_type == 'super':
                    base_dir = os.path.dirname(target_dir) if target_dir != root_dir else target_dir
                else:
                    base_dir = target_dir

                cand1 = os.path.join(base_dir, f"{subpath}.rs")
                cand2 = os.path.join(base_dir, subpath, "mod.rs")
                cand3 = os.path.join(base_dir, f"{subpath}/mod.rs")

                if not check_and_add(cand1) and not check_and_add(cand2):
                    check_and_add(cand3)

        return found_rel_paths

File: codecontext_core/src/services/dependency_service.rs
use regex::Regex;
use std::collections::HashSet;
use std::path::{Path, PathBuf};

pub fn trace_dependencies(root_dir: &str, target_rel_path: &str, content: &str) -> HashSet<String> {
    let mut found = HashSet::new();
    if root_dir.is_empty() || target_rel_path.is_empty() || content.is_empty() {
        return found;
    }

    let ext = Path::new(target_rel_path)
        .extension()
        .and_then(|s| s.to_str())
        .unwrap_or("")
        .to_lowercase();

    let full_target_path = Path::new(root_dir).join(target_rel_path);
    let target_dir = full_target_path.parent().unwrap_or_else(|| Path::new(root_dir));

    let mut check_and_insert = |path: PathBuf| -> bool {
        if path.is_file() {
            if let Ok(rel) = path.strip_prefix(root_dir) {
                found.insert(rel.to_string_lossy().replace('\\', "/"));
                return true;
            }
        }
        false
    };

    match ext.as_str() {
        "py" => {
            let import_re = Regex::new(r"(?m)^\s*import\s+([a-zA-Z0-9_.,\s]+)").unwrap();
            let from_re = Regex::new(r"(?m)^\s*from\s+(\.?\.?[a-zA-Z0-9_.]+)\s+import\s+([a-zA-Z0-9_.,\s()]+)").unwrap();

            let mut targets = Vec::new();

            for caps in import_re.captures_iter(content) {
                if let Some(mat) = caps.get(1) {
                    for part in mat.as_str().split(',') {
                        let clean = part.trim();
                        if !clean.is_empty() {
                            targets.push((String::new(), clean.to_string()));
                        }
                    }
                }
            }

            for caps in from_re.captures_iter(content) {
                let base = caps.get(1).map_or("", |m| m.as_str()).to_string();
                if let Some(imports_mat) = caps.get(2) {
                    let clean_imports = imports_mat.as_str()
                        .replace('(', "")
                        .replace(')', "")
                        .replace('\n', " ")
                        .replace('\r', " ");
                    for part in clean_imports.split(',') {
                        let clean_item = part.trim();
                        if !clean_item.is_empty() {
                            targets.push((base.clone(), clean_item.to_string()));
                        }
                    }
                }
            }

            for (base, item) in targets {
                let has_base = !base.is_empty();
                let full_import_path = if has_base {
                    format!("{}.{}", base, item)
                } else {
                    item.clone()
                };

                let dot_count = if has_base {
                    base.chars().take_while(|c| *c == '.').count()
                } else {
                    item.chars().take_while(|c| *c == '.').count()
                };

                let clean_path = full_import_path.trim_start_matches('.');
                let subpath = clean_path.replace('.', "/");

                let search_bases: Vec<PathBuf> = if dot_count > 0 {
                    let mut temp_base = target_dir.to_path_buf();
                    for _ in 0..(dot_count - 1) {
                        if let Some(parent) = temp_base.parent() {
                            temp_base = parent.to_path_buf();
                        }
                    }
                    vec![temp_base]
                } else {
                    vec![
                        Path::new(root_dir).to_path_buf(),
                        Path::new(root_dir).join("src"),
                        target_dir.to_path_buf(),
                    ]
                };

                for base_dir in search_bases {
                    if !subpath.is_empty() {
                        let cand1 = base_dir.join(format!("{}.py", subpath));
                        let cand2 = base_dir.join(&subpath).join("__init__.py");
                        if check_and_insert(cand1) || check_and_insert(cand2) {
                            break;
                        }
                    }
                    if has_base {
                        let base_subpath = base.trim_start_matches('.').replace('.', "/");
                        if !base_subpath.is_empty() {
                            let cand1 = base_dir.join(format!("{}.py", base_subpath));
                            let cand2 = base_dir.join(&base_subpath).join("__init__.py");
                            if check_and_insert(cand1) || check_and_insert(cand2) {
                                break;
                            }
                        }
                    }
                }
            }
        }
        "js" | "jsx" | "ts" | "tsx" => {
            let es6_re = Regex::new(r#"(?s)\bimport\s+[^"']*?\s+from\s+["'](\..*?)["']"#).unwrap();
            let dynamic_re = Regex::new(r#"\bimport\(\s*["'](\..*?)["']\s*\)"#).unwrap();
            let require_re = Regex::new(r#"\brequire\(\s*["'](\..*?)["']\s*\)"#).unwrap();

            let mut imports = Vec::new();
            for caps in es6_re.captures_iter(content) {
                if let Some(m) = caps.get(1) {
                    imports.push(m.as_str());
                }
            }
            for caps in dynamic_re.captures_iter(content) {
                if let Some(m) = caps.get(1) {
                    imports.push(m.as_str());
                }
            }
            for caps in require_re.captures_iter(content) {
                if let Some(m) = caps.get(1) {
                    imports.push(m.as_str());
                }
            }

            for rel_import in imports {
                let abs_base = target_dir.join(rel_import);
                let possible_exts = ["", ".js", ".ts", ".jsx", ".tsx", "/index.js", "/index.ts"];

                for p_ext in possible_exts {
                    let test_path = PathBuf::from(format!("{}{}", abs_base.to_string_lossy(), p_ext));
                    if check_and_insert(test_path) {
                        break;
                    }
                }
            }
        }
        "c" | "cpp" | "h" | "hpp" => {
            let include_re = Regex::new(r#"(?m)^\s*#include\s+["']([^"']+)["']"#).unwrap();
            for caps in include_re.captures_iter(content) {
                if let Some(mat) = caps.get(1) {
                    let header = mat.as_str();
                    let test_path1 = target_dir.join(header);
                    let test_path2 = Path::new(root_dir).join(header);

                    if !check_and_insert(test_path1) {
                        check_and_insert(test_path2);
                    }
                }
            }
        }
        "rs" => {
            let mod_re = Regex::new(r"(?m)^\s*(?:pub\s+)?mod\s+([a-zA-Z0-9_]+);").unwrap();
            let use_re = Regex::new(r#"(?m)^\s*(?:pub\s+)?use\s+(crate|super|self)::([a-zA-Z0-9_:]+)"#).unwrap();

            for caps in mod_re.captures_iter(content) {
                if let Some(mat) = caps.get(1) {
                    let mod_name = mat.as_str();
                    let cand1 = target_dir.join(format!("{}.rs", mod_name));
                    let cand2 = target_dir.join(mod_name).join("mod.rs");

                    if !check_and_insert(cand1) {
                        check_and_insert(cand2);
                    }
                }
            }

            for caps in use_re.captures_iter(content) {
                let base_type = caps.get(1).map_or("", |m| m.as_str());
                if let Some(path_mat) = caps.get(2) {
                    let raw_path = path_mat.as_str();
                    let subpath = raw_path.replace("::", "/");

                    let base_dir = match base_type {
                        "crate" => {
                            let src_dir = Path::new(root_dir).join("src");
                            if src_dir.is_dir() {
                                src_dir
                            } else {
                                Path::new(root_dir).to_path_buf()
                            }
                        }
                        "super" => target_dir.parent().unwrap_or(target_dir).to_path_buf(),
                        _ => target_dir.to_path_buf(),
                    };

                    let cand1 = base_dir.join(format!("{}.rs", subpath));
                    let cand2 = base_dir.join(&subpath).join("mod.rs");
                    let cand3 = base_dir.join(format!("{}/mod.rs", subpath));

                    if !check_and_insert(cand1) && !check_and_insert(cand2) {
                        check_and_insert(cand3);
                    }
                }
            }
        }
        _ => {}
    }

    found
}